File: version_1_5_Julia/models/fsa_high_res/_plant.jl
# FSA v1.5 plant — purely functional, stateless.
#
# An immutable PlantState record + two pure functions:
#   plant_step(s, Φ_t, params, dt, key)        → (state, obs)
#   plant_rollout(s0, Φ_subdaily, params, dt, key0) → trajectory + obs vectors
#
# No mutable struct, no `!`-suffix, no MersenneTwister threaded as state,
# no history dict. The bench builds up a history by collecting returned
# values via foldl / accumulate (see bench_smc_full_mpc_fsa_gpu.jl).

module Plant

using StaticArrays
using StableRNGs
using Match

import ..Dynamics: drift, diffusion_state_dep
import ..Simulation: sample_obs_bfa, DEFAULT_PARAMS, INIT_STATE, DT_BIN_DAYS,
                     params_v15_to_v1_nt

export PlantState, plant_step, plant_rollout, init_plant_state


# ── Immutable state record ────────────────────────────────────────────────

"""
    PlantState

Immutable plant state. (B, F, A) carried as an SVector for stack
allocation; t_bin is the global bin counter (advances by 1 per
`plant_step`).
"""
struct PlantState
    bfa::SVector{3, Float64}
    t_bin::Int
end

"""
    init_plant_state(; init_state = INIT_STATE) -> PlantState

Build a fresh PlantState at the canonical init.
"""
function init_plant_state(; init_state = INIT_STATE)
    return PlantState(
        SVector{3, Float64}(Float64(init_state.B),
                            Float64(init_state.F),
                            Float64(init_state.A)),
        0,
    )
end


# ── Boundary reflection helper (pure) ─────────────────────────────────────
# Implemented via `@match` with guards — maps 1:1 to the Lean4 port's
# `match x with | x => if x < 0 then -x else if x > 1 then 2 - x else x`.

@inline _reflect_unit(x::Float64) = @match x begin
    x, if x < 0.0 end => -x
    x, if x > 1.0 end => 2.0 - x
    _                 => x
end


# ── Pure single-bin step ──────────────────────────────────────────────────

"""
    plant_step(s, Φ_t, params, dt, key) -> NamedTuple

One Euler-Maruyama step under control `Φ_t` from state `s`. Returns

    (state = PlantState(...), obs = (obs_B, obs_F, obs_A))

Same `(s, Φ_t, params, dt, key)` always returns the same result.
"""
function plant_step(s::PlantState,
                     Φ_t::Float64,
                     params::Dict{Symbol, Float64},
                     dt::Float64,
                     key::UInt64)
    rng = StableRNG(key)

    # v1.5 → v1 basis adapter: drift expects (kappa_B, kappa_F, ...).
    params_nt = params_v15_to_v1_nt(params)

    bfa = s.bfa
    d   = drift(bfa, params_nt, Φ_t)              # Vector{Float64}, length 3
    σ   = diffusion_state_dep(bfa, params_nt)     # Vector{Float64}, length 3
    ξ   = randn(rng, 3)                            # Vector{Float64}, length 3

    # SDE step (Float64 throughout — no fp32 inside the plant)
    y_pred = bfa .+ d .* dt .+ σ .* sqrt(dt) .* SVector{3, Float64}(ξ[1], ξ[2], ξ[3])

    # Boundary reflection (B ∈ [0, 1], F & A ≥ 0)
    y_next = SVector{3, Float64}(
        _reflect_unit(y_pred[1]),
        abs(y_pred[2]),
        abs(y_pred[3]),
    )

    # Obs from new state — independent key derived from input key
    obs = sample_obs_bfa(y_next, params, hash((key, :obs)))

    return (
        state = PlantState(y_next, s.t_bin + 1),
        obs   = obs,
    )
end


# ── Pure stride rollout ───────────────────────────────────────────────────

"""
    plant_rollout(s0, Φ_subdaily, params, dt, key0) -> NamedTuple

Apply `length(Φ_subdaily)` consecutive `plant_step`s from `s0`. Returns

    (final_state = PlantState(...),
     trajectory  = Matrix{Float64} of shape (stride_bins, 3),
     obs_B, obs_F, obs_A = Vector{Float32} of length stride_bins,
     Phi          = Vector{Float32} of length stride_bins)

The function is pure: same inputs always produce the same outputs.
"""
function plant_rollout(s0::PlantState,
                        Φ_subdaily::AbstractVector,
                        params::Dict{Symbol, Float64},
                        dt::Float64,
                        key0::UInt64)
    n     = length(Φ_subdaily)
    traj  = Matrix{Float64}(undef, n, 3)
    obs_B = Vector{Float64}(undef, n)
    obs_F = Vector{Float64}(undef, n)
    obs_A = Vector{Float64}(undef, n)

    s = s0
    @inbounds for k in 1:n
        sub_key = hash((key0, :step, k))
        nxt     = plant_step(s, Float64(Φ_subdaily[k]), params, dt, sub_key)
        s       = nxt.state
        traj[k, 1] = s.bfa[1]
        traj[k, 2] = s.bfa[2]
        traj[k, 3] = s.bfa[3]
        obs_B[k]   = nxt.obs.obs_B
        obs_F[k]   = nxt.obs.obs_F
        obs_A[k]   = nxt.obs.obs_A
    end

    return (
        final_state = s,
        trajectory  = traj,
        obs_B = Float32.(obs_B),
        obs_F = Float32.(obs_F),
        obs_A = Float32.(obs_A),
        Phi   = Float32.(Φ_subdaily),
    )
end

end # module Plant

File: version_1_5_Python_JAX/models/fsa_high_res/_plant.py
"""FSA v1.5 plant — purely-functional, stateless. Direct port of
`version_1_5_Julia/models/fsa_high_res/_plant.jl`.

Two pure functions over an immutable (B, F, A, t_bin) state record:

  plant_step(state, Phi_t, params, dt, sde_noise, obs_noise)
      → (next_state, obs)

  plant_rollout(s0, Phi_subdaily, params, dt, sde_noise_seq, obs_noise_seq)
      → (final_state, traj, obs_B, obs_F, obs_A, Phi)

No mutable class, no `!`-suffix, no RNG threaded as state. The bench
builds up a history by collecting returned values via accumulator.

RNG handling: the Julia source draws 3 SDE noise + 3 obs noise per
bin from `StableRNG(key)`. The Python+JAX equivalent takes pre-drawn
noise tuples directly so:
  (a) the diff test against Julia is straightforward (same noise → same
      output, no need to reimplement Julia's hash-based key derivation),
  (b) JAX's pure-RNG style works naturally — caller splits the key.

`PlantState` is a tiny dataclass-like NamedTuple for clarity; mirrors
the Julia `struct PlantState { bfa::SVector{3,Float64}, t_bin::Int }`.
"""

from __future__ import annotations

from typing import NamedTuple

import jax
import jax.numpy as jnp
import numpy as np

from models.fsa_high_res._dynamics import drift_jax, diffusion_state_dep
from models.fsa_high_res.simulation import (
    INIT_STATE, params_v15_to_v1, sample_obs_bfa,
)


class PlantState(NamedTuple):
    """Immutable plant state. (B, F, A) carried as a length-3 jnp.ndarray;
    t_bin is the global bin counter (advances by 1 per `plant_step`).
    Matches the Julia `struct PlantState` at `_plant.jl:33-36`.
    """
    bfa: jnp.ndarray   # shape (3,) float64
    t_bin: int


def init_plant_state(init_state=None) -> PlantState:
    """Build a fresh PlantState at the canonical init. Matches
    `_plant.jl:43-50` (`init_plant_state`).
    """
    s = init_state if init_state is not None else INIT_STATE
    return PlantState(
        bfa=jnp.array([float(s['B']), float(s['F']), float(s['A'])],
                      dtype=jnp.float64),
        t_bin=0,
    )


def _reflect_unit(x):
    """★ @match site #2 from writeup §7.1 (Julia `_plant.jl:57-61`).
    The Julia `@match x begin x, if x < 0 end => -x; … end` is
    `if/else if/else` here — same evaluation semantics on Float.
    """
    return jnp.where(x < 0.0, -x, jnp.where(x > 1.0, 2.0 - x, x))


def plant_step(state: PlantState, Phi_t: float, params: dict,
                dt: float, sde_noise, obs_noise=None,
                obs_noise_key=None) -> tuple:
    """One Euler-Maruyama step under control `Phi_t` from `state`.

    Args:
        state: PlantState (immutable).
        Phi_t: scalar — control input at this bin.
        params: v1.5-form dict with full 14 dynamics + 3 obs noise keys.
        dt: bin width in days.
        sde_noise: shape (3,) array — pre-drawn standard-normal triple
            for the SDE step.
        obs_noise: shape (3,) array — pre-drawn standard-normal triple
            for the obs sample. If None, `obs_noise_key` (a JAX PRNGKey)
            is used to draw it via `sample_obs_bfa`.
        obs_noise_key: optional JAX PRNGKey used to sample obs noise
            when `obs_noise` is None.

    Returns:
        (PlantState, dict) — next state and `(obs_B, obs_F, obs_A)`.
    """
    params_v1 = params_v15_to_v1(params)
    bfa = state.bfa
    d = drift_jax(bfa, params_v1, Phi_t)
    sigma = diffusion_state_dep(bfa, params_v1)
    noise = jnp.asarray(sde_noise, dtype=jnp.float64)
    y_pred = bfa + d * dt + sigma * jnp.sqrt(dt) * noise

    # Boundary reflection (B ∈ [0, 1], F & A ≥ 0)
    next_bfa = jnp.array([
        _reflect_unit(y_pred[0]),
        jnp.abs(y_pred[1]),
        jnp.abs(y_pred[2]),
    ])

    next_state = PlantState(bfa=next_bfa, t_bin=state.t_bin + 1)

    # Obs sample. Two paths kept for symmetry with the Julia API:
    # (a) caller supplies obs_noise — bit-identical to a Julia driver
    #     that pre-draws the noise. Used by the diff test.
    # (b) caller supplies obs_noise_key — JAX-native sampling for the
    #     bench / closed-loop path.
    if obs_noise is not None:
        obs = dict(
            obs_B=float(next_bfa[0]) + params['sigma_B_obs'] * float(obs_noise[0]),
            obs_F=float(next_bfa[1]) + params['sigma_F_obs'] * float(obs_noise[1]),
            obs_A=float(next_bfa[2]) + params['sigma_A_obs'] * float(obs_noise[2]),
        )
    else:
        if obs_noise_key is None:
            raise ValueError("plant_step needs obs_noise or obs_noise_key")
        obs = sample_obs_bfa(next_bfa, params, obs_noise_key)

    return next_state, obs


def plant_rollout(s0: PlantState, Phi_subdaily, params: dict,
                   dt: float, key0):
    """Apply `len(Phi_subdaily)` consecutive `plant_step`s from `s0`.
    Matches `_plant.jl:124-156` (`plant_rollout`).

    Args:
        s0: initial PlantState.
        Phi_subdaily: shape (n,) — per-bin Φ values.
        params: v1.5-form dict (14 dynamics + 3 obs noise).
        dt: bin width in days.
        key0: JAX PRNGKey; sub-keys for each bin's SDE + obs sampling
            are split deterministically from this.

    Returns:
        dict with keys:
            final_state: PlantState
            trajectory:  ndarray (n, 3)
            obs_B, obs_F, obs_A: ndarray (n,)
            Phi:         ndarray (n,)
    """
    n = int(len(Phi_subdaily))
    Phi_arr = np.asarray(Phi_subdaily, dtype=np.float64)
    traj = np.zeros((n, 3), dtype=np.float64)
    obs_B = np.zeros(n, dtype=np.float64)
    obs_F = np.zeros(n, dtype=np.float64)
    obs_A = np.zeros(n, dtype=np.float64)

    s = s0
    key = key0
    for k in range(n):
        key, sde_key, obs_key = jax.random.split(key, 3)
        sde_noise = jax.random.normal(sde_key, (3,), dtype=jnp.float64)
        s_next, obs = plant_step(
            s, float(Phi_arr[k]), params, dt,
            sde_noise=sde_noise, obs_noise_key=obs_key,
        )
        s = s_next
        traj[k, 0] = float(s.bfa[0])
        traj[k, 1] = float(s.bfa[1])
        traj[k, 2] = float(s.bfa[2])
        obs_B[k] = obs['obs_B']
        obs_F[k] = obs['obs_F']
        obs_A[k] = obs['obs_A']

    return dict(
        final_state=s,
        trajectory=traj,
        obs_B=obs_B,
        obs_F=obs_F,
        obs_A=obs_A,
        Phi=Phi_arr,
    )
